### ISSUE TITLE
Sprockets 3 errors with premailer

### ISSUE THREAD

**Issue by marclennox** [fphilipe/premailer-rails#137]
I'm getting the following error in premailer with the recent Sprockets update to 3.x

```
NoMethodError: undefined method `split' for nil:NilClass
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_loaders/network_loader.rb:17:in `uri_for_url'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_loaders/network_loader.rb:8:in `load'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_helper.rb:33:in `block in load_css'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_helper.rb:32:in `each'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_helper.rb:32:in `load_css'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_helper.rb:19:in `block in css_for_doc'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_helper.rb:19:in `map'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/css_helper.rb:19:in `css_for_doc'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/customized_premailer.rb:15:in `initialize'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:85:in `new'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:85:in `premailer'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:75:in `generate_html_part'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:52:in `generate_html_part_replacement'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:24:in `perform'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/premailer-rails-1.8.0/lib/premailer/rails/hook.rb:8:in `perform'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/mail-2.6.3/lib/mail/mail.rb:242:in `block in inform_interceptors'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/mail-2.6.3/lib/mail/mail.rb:241:in `each'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/mail-2.6.3/lib/mail/mail.rb:241:in `inform_interceptors'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/mail-2.6.3/lib/mail/message.rb:224:in `inform_interceptors'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/mail-2.6.3/lib/mail/message.rb:234:in `deliver'
	from /usr/local/lib/bundle/ruby/2.2.0/gems/actionmailer-4.2.1/lib/action_mailer/message_delivery.rb:85:in `deliver_now'
```

**Comment by fphilipe** [comment]
Haven't played with sprockets 3 yet. I'll take a closer look at it.

**Comment by fphilipe** [comment]
What's your `asset_host` config in your rails app?

**Comment by marclennox** [comment]
nil.  I don't use a separate asset host.

**Comment by marclennox** [comment]
I had a look at some issues being raised on the sprockets project, and it seems there are other things breaking for people with the sprockets 3 release.

**Comment by rywall** [comment]
This is also happening for me. My asset_host is also nil.

**Comment by daniel-rikowski** [comment]
It looks like `AssetPipelineLoader` fails with Sprockets 3.0 and then `NetworkLoader` is tried.

`NetworkLoader` itself seems to be very brittle, it fails if there is no `asset_host` configured or if the asset host has also a port  (That's what brought me here)

**Comment by daniel-rikowski** [comment]
**tl;dr**

Sprockets 3 has longer hashes in file names, which are not detected by premailer-rails. Monkey-patch is trivial.

**Full version:**

The problem is that Sprockets 3 generates a longer hash after the logical asset name than Sprockets 2:

Sprockets 2: 
`mailer.css` becomes `mailer-01234567890123456789012345678912.css`

Sprockets 3:
`mailer.css` becomes `mailer-0123456789012345678901234567890123456789012345678901234567891234.css`

Premailer-rails looks for *exactly* 32 characters and fails to strip the longer hash. Subsequently Sprockets can no longer find the asset file on disk. (As you have to ask for a logical asset name, i.e. "mailer.css" and not "mailer-xxx.css")

One solution would be to not look for exactly 32 characters but for *at least* 32 characters.

In premailer/rails/css_loaders/asset_pipeline_loader.rb: 

Replace:
```ruby
        def file_name(url)
          URI(url).path
            .sub("#{::Rails.configuration.assets.prefix}/", '')
            .sub(/-\h{32}\.css$/, '.css')
        end
```
With: 
```ruby
        def file_name(url)
          URI(url).path
            .sub("#{::Rails.configuration.assets.prefix}/", '')
            .sub(/-\h{32,}\.css$/, '.css') 
        end
```
(Notice the comma in `\h{32,}`)


I also looked at NetworkLoader. For Sprockets 3 nothing needs to be fixed here, but it has a lot of other issues. It can't imagine that it ever worked for asset URLs.

First of all replacing the scheme of a generic URL is allowed but it does not change the underlying instance class. (i.e. a `URI::Generic` does not change into a `URI::HTTP` if the scheme is assigned) But `Net::HTTP.get` calls the `request_uri` method which is not available in `URI::Generic` but in `URI::HTTP`.
Also there is some manual URI parsing which leads to the port bug I mentioned above.

In premailer/rails/css_loaders/network_loader.rb: 

By replacing
```ruby
def uri_for_url(url)
  uri = URI(url)
  if not valid_uri?(uri) and defined?(::Rails)
	scheme, host =
	  ::Rails.configuration.action_controller.asset_host.split(%r{:?//})
	scheme = 'http' if scheme.blank?
	uri.scheme ||= scheme
	uri.host ||= host
  end

  uri if valid_uri?(uri)
end
```
with: 
```ruby
def uri_for_url(url)
  uri = URI(url)
  if not valid_uri?(uri) and defined?(::Rails)
	asset_uri = URI(::Rails.configuration.action_controller.asset_host) rescue return
	uri.host ||= asset_uri.host
	uri.port ||= asset_uri.port
	unless uri.scheme
	  uri.scheme = asset_uri.scheme || 'http'
	  uri = URI(uri.to_s)
	end
  end

  uri if valid_uri?(uri)
end
```
the overall quality of the code could be increased a lot.

**Comment by mfkp** [comment]
Just ran into this issue as well, my asset_host is nil and it's failing in the same way as above.
